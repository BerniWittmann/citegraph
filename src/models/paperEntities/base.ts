import { AuthorFields } from '@/models/paperEntities/author'
import { ConferenceFields } from '@/models/paperEntities/conference'
import { CountryFields } from '@/models/paperEntities/country'
import { InstitutionFields } from '@/models/paperEntities/institution'
import { PublicationFields } from '@/models/paperEntities/publication'
import { PublisherFields } from '@/models/paperEntities/publisher'
import { RecordFields } from '@/models/paperEntities/record'
import { constructQuery, PaperEntityQueryParameters } from '@/models/paperEntities/query'
import { constructMutation, PaperEntityMutationParameters } from '@/models/paperEntities/mutation'
import Transformer from '@/transformers/Transformer'

export interface PaperEntityBaseFields {
  id: string;
}

export default abstract class PaperEntity implements PaperEntityBaseFields {
  id: string;
  static readonly key: string;
  static readonly schemaName: string;
  static readonly queryName: string;
  static readonly displayedColumns: Array<PaperEntityTableColumn>;
  static readonly queryFields: string;
  static readonly mutationFields: Array<string> = [];
  static readonly transformer: typeof Transformer = Transformer;

  protected constructor ({ id }: PaperEntityBaseFields) {
    this.id = id
  }

  static getQuery (params: PaperEntityQueryParameters): string {
    return constructQuery(this.queryName, this.schemaName, params, this.queryFields)
  }

  static getMutation (params: PaperEntityMutationParameters, data: Partial<PaperEntityFields>): string {
    return constructMutation(this.queryName, this.schemaName, params, this.mutationFields, data)
  }
}

export interface PaperEntityTableColumn {
  text: string;
  value: string;
  align?: 'start' | 'center' | 'end'
  sortable?: boolean
  filterable?: boolean
  divider?: boolean
  class?: string | string[]
  width?: string | number
  filter?: (value: any, search: string, item: any) => boolean
  sort?: (a: any, b: any) => number,
  displayedByDefault?: boolean
}

export type PaperEntityFields = PaperEntityBaseFields & (
  AuthorFields &
  ConferenceFields &
  CountryFields &
  InstitutionFields &
  PublicationFields &
  PublisherFields &
  RecordFields
)
