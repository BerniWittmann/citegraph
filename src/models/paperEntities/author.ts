import PaperEntity, { PaperEntityBaseFields, PaperEntityTableColumn } from './base'
import AuthorTransformer from '@/transformers/AuthorTransformer'

export interface AuthorFields {
  firstName: string
  lastName: string
  countRecords: number
}

export default class Author extends PaperEntity implements AuthorFields {
  static key: string = 'author'
  static schemaName: string = 'author'
  static queryName: string = 'Authors'
  static queryFields: string = `id,
          first_name,
          last_name,
          number_citations`
  static mutationFields: Array<string> = ['first_name', 'last_name']
  static displayedColumns: Array<PaperEntityTableColumn> = [{
    text: 'project.explore.table.headers.author.first_name',
    value: 'firstName',
    sortable: true,
    filterable: true,
    displayedByDefault: true
  }, {
    text: 'project.explore.table.headers.author.last_name',
    value: 'lastName',
    sortable: true,
    filterable: true,
    displayedByDefault: true
  }, {
    text: 'project.explore.table.headers.author.count_records',
    value: 'countRecords',
    sortable: true,
    align: 'center'
  }]
  static transformer = AuthorTransformer
  firstName: string
  lastName: string
  countRecords: number

  constructor ({ id, firstName, lastName, countRecords }: PaperEntityBaseFields & AuthorFields) {
    super({ id })
    this.firstName = firstName
    this.lastName = lastName
    this.countRecords = countRecords
  }
}
