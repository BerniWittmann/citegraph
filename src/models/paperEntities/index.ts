import Author from './author'
import Conference from './conference'
import Country from './country'
import Institution from './institution'
import Publication from './publication'
import Publisher from './publisher'
import Record from './record'
import PaperEntity, { PaperEntityTableColumn, PaperEntityFields } from './base'

export const entities = [
  Record,
  Author,
  Institution,
  Country,
  Publication,
  Publisher,
  Conference
]

export const entityKeysMap = {
  [Record.key]: Record,
  [Author.key]: Author,
  [Institution.key]: Institution,
  [Country.key]: Country,
  [Publication.key]: Publication,
  [Publisher.key]: Publisher,
  [Conference.key]: Conference
}

export const entityKeys: Array<string> = entities.map((entity) => entity.key)

export type PaperEntityTableColumn = PaperEntityTableColumn

export type PaperEntity = PaperEntity

export type PaperEntityFields = PaperEntityFields

export default {
  Author,
  Conference,
  Country,
  Institution,
  Publication,
  Publisher,
  Record
}
