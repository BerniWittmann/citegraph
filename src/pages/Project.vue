<template>
  <div class="project">
    <v-alert v-if="!hasProject" type="error" prominent>
      {{ $t('project.not_available') }}
    </v-alert>

    <div v-if="hasProject">
      <h1>{{ project.name }}</h1>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project from '@/models/project'

@Component
export default class ProjectPage extends Vue {
  get project (): Project | undefined {
    return this.$store.getters['projects/activeProject']
  }

  get hasProject (): boolean {
    return this.$store.getters['projects/hasActiveProject']
  }
}
</script>
