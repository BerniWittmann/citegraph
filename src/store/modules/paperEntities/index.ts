import { Module } from 'vuex'
import { PaperEntitiesState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const paperEntities: Module<PaperEntitiesState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
