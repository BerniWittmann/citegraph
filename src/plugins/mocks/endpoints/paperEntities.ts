import db from '../db'
import MockAdapter from 'axios-mock-adapter'
import { RecordFields } from '@/models/paperEntities/record'
import { setupMocks as setupGraphQLMocks } from './graphql-helpers/endpoint'
import { QueryConfiguration } from './graphql-helpers/queries'
import { AuthorFields } from '@/models/paperEntities/author'
import { CountryFields } from '@/models/paperEntities/country'

const records = db.getCollection('records')
const authors = db.getCollection('authors')
const countries = db.getCollection('countries')

const queryResponseConfiguration: Record<string, QueryConfiguration> = {
  'Records': {
    schemaName: 'record',
    queryName: 'Records',
    collection: records,
    filterFunction: recordsFilter
  },
  'Authors': {
    schemaName: 'author',
    queryName: 'Authors',
    collection: authors,
    filterFunction: authorsFilter
  },
  'Countries': {
    schemaName: 'country',
    queryName: 'Countries',
    collection: countries,
    filterFunction: countriesFilter
  }
}

function recordsFilter (obj: RecordFields, filter: string) {
  return obj.title.toLowerCase().includes(filter) ||
    obj.year.toString().includes(filter) ||
    obj.keywords.map(keyword => keyword.toLowerCase()).join('').includes(filter) ||
    obj.authors.map(author => (author.firstName + ' ' + author.lastName).toLowerCase()).join(';').includes(filter)
}

function authorsFilter (obj: AuthorFields, filter: string) {
  return obj.firstName.toLowerCase().includes(filter) ||
    obj.lastName.toLowerCase().includes(filter)
}

function countriesFilter (obj: CountryFields, filter: string) {
  return obj.name.toLowerCase().includes(filter)
}

export function setupMocks (mock: MockAdapter) {
  setupGraphQLMocks(mock, queryResponseConfiguration)
}
