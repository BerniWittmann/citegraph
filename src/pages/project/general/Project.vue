import { ProjectStates } from '@/models/project'
<template>
  <div class="project">
    <div class="d-flex mr-5 mt-4">
      <h1>{{ project.name }}</h1>
      <v-spacer/>
      <v-btn :to="{ name: 'projects.single.edit', params: { projectId: project.id } }">
        <v-icon>mdi-pencil</v-icon>
        {{ $t('project.edit.button_text') }}
      </v-btn>
    </div>
    <v-container v-if="hasNoDataYet">
      <v-row justify="center">
        <v-col :lg="3" class="d-flex justify-center flex-column text-center">
          <empty-icon class="align-self-center mb-6"></empty-icon>
          <h3 class="headline">{{ $t('project.single.no_data.description') }}</h3>
          <p>{{ $t('project.single.no_data.import_data') }}</p>
          <v-btn color="primary" rounded large :to="{ name: 'projects.single.import', params: { projectId: project.id } }">
            {{ $t('project.single.no_data.button') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-container>
    <v-container v-else>
      <v-row v-if="datasetInformation">
        <v-col>
          <v-card>
            <v-card-title>{{ $t('project.single.dataset.title') }}</v-card-title>
            <v-card-text>
              <v-container>
                <v-row>
                  <v-col v-for="(count, name) in datasetInformation" :key="name">
                    <v-card outlined class="d-flex flex-column justify-center text-center pa-5">
                      <p class="headline">{{count}}</p>
                      <p class="mb-0">{{ $t('project.single.dataset.' + name ) }}</p>
                    </v-card>
                  </v-col>
                </v-row>
              </v-container>
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
      <v-row>
        <v-col lg="3" md="6" cols="12" v-if="project.topAuthors">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_authors') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="author in project.topAuthors" :key="author.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'author',
                 entityId: author.id
                }}">
                  {{ author.firstName }} {{ author.lastName }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.topRecords">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_records') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="record in project.topRecords" :key="record.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'record',
                 entityId: record.id
                }}">
                  {{ record.title }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.countryRelevanceData">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.country_relevance') }}
            </v-card-title>
            <v-card-text>
              TODO Diagram here
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12">
          <v-row no-gutters>
            <v-col v-if="project.topKeywords" class="mb-6">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.top_10_keywords') }}
                </v-card-title>
                <v-card-text>
                  <v-chip-group multiple column>
                    <v-chip v-for="keyword in project.topKeywords" :key="keyword">{{ keyword }}</v-chip>
                  </v-chip-group>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
          <v-row no-gutters>
            <v-col v-if="project.counts">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.visualizations') }}
                </v-card-title>
                <v-card-text>
                  <div class="d-flex flex-column justify-center text-center pa-5" v-if="hasNoVisualizationsYet">
                    <p class="mb-0">{{ $t('project.single.overview.no_visualizations_yet') }}</p>
                    <v-btn color="primary" rounded>{{ $t('project.single.overview.add_visualization') }}</v-btn>
                  </div>
                  <v-card outlined class="d-flex flex-column justify-center text-center pa-5" v-else>
                    <p class="headline">{{ project.counts.visualizations }}</p>
                    <p class="mb-0">{{ $t('project.single.overview.visualizations') }}</p>
                  </v-card>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project, { ProjectStates } from '@/models/project'
import EmptyIcon from '@/assets/svgs/Empty.vue'

@Component({
  components: {
    EmptyIcon
  }
})
export default class ProjectPage extends Vue {
  get project (): Project {
    return this.$store.getters['projects/activeProject']
  }

  get datasetInformation (): Object | undefined {
    if (!this.project.counts) return undefined
    const data = { ...this.project.counts }
    delete data['visualizations']
    return data
  }

  get hasNoDataYet (): boolean {
    return this.project.state === ProjectStates.IMPORT
  }

  get hasNoVisualizationsYet (): boolean {
    return this.project.state === ProjectStates.EXPLORE
  }
}
</script>
