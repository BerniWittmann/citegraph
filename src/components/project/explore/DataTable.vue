<template>
  <v-card style="width: 100%">
    <v-card-title>
      <h1 class="display-1 my-5">
        {{ $t('project.explore.title.query_by_type.' + queryByType ) }}
      </h1>
      <v-spacer></v-spacer>
      <v-text-field
        v-model="currentOptions.search"
        append-icon="mdi-magnify"
        :label="$t('project.explore.table.search' )"
        single-line
        hide-details
        clearable
      ></v-text-field>
    </v-card-title>
    <v-data-table
      :loading="loading"
      style="width: 100%"
      :headers="headers"
      :items="items"
      :page.sync="currentOptions.page"
      :items-per-page.sync="currentOptions.itemsPerPage"
      @update:options="updateOptions"
      disable-filtering
      :server-items-length="totalItemCount"
    ></v-data-table>
  </v-card>
</template>

<script lang="ts">
import { Component, Prop, Vue, Watch } from 'vue-property-decorator'
import { entityKeysMap, PaperEntity, PaperEntityTableColumn } from '@/models/paperEntities'
import { debounce } from 'debounce'
import Project from '@/models/project'
import deepEqual from 'deep-equal'

type UpdateOptions = {
  page: number
  itemsPerPage: number
  sortBy: string[]
  sortDesc: boolean[],
  search?: string
}

@Component
export default class DataTable extends Vue {
  loading: boolean = false
  currentOptions: UpdateOptions = {
    page: 1,
    itemsPerPage: 10,
    sortBy: [],
    sortDesc: [],
    search: ''
  }

  @Prop(String) readonly queryByType!: string

  get entityClass () {
    return entityKeysMap[this.queryByType]
  }

  get items (): Array<PaperEntity> {
    return this.$store.getters['paperEntities/entities']
  }

  get headers (): Array<PaperEntityTableColumn> {
    return this.entityClass.displayedColumns.map((header) => ({
      ...header,
      text: this.$t(header.text).toString()
    }))
  }

  get project (): Project {
    return this.$store.getters['projects/activeProject']
  }

  get totalItemCount (): number {
    return this.$store.getters['paperEntities/entityCount']
  }

  async reloadData (): Promise<void> {
    this.loading = true
    await this.$store.dispatch('paperEntities/fetchEntities', {
      projectId: this.project.id,
      entityType: this.queryByType,
      perPage: this.currentOptions.itemsPerPage,
      pageOffset: this.currentOptions.page - 1,
      filter: this.currentOptions.search
    })
    this.loading = false
  }

  get debouncedReloadData (): Function {
    return debounce(this.reloadData, 300)
  }

  @Watch('currentOptions.search')
  handleSearch (newVal: string): void {
    if (newVal === this.currentOptions.search) return
    this.loading = true
    this.currentOptions.search = newVal
    this.currentOptions.page = 1
    this.debouncedReloadData()
  }

  updateOptions (options: UpdateOptions): void {
    if (deepEqual(options, this.currentOptions)) return
    this.loading = true
    this.currentOptions = {
      ...this.currentOptions,
      ...options
    }
    this.debouncedReloadData()
  }

  @Watch('queryByType')
  updateQueryByType (): void {
    this.currentOptions.page = 1
    this.currentOptions.itemsPerPage = 10
    this.currentOptions.search = ''
    this.currentOptions.sortBy = []
    this.currentOptions.sortDesc = []
    this.reloadData()
  }
}
</script>
