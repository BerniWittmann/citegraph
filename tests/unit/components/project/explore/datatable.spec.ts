import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../setupPlugins'

import DataTable from '@/components/project/explore/DataTable.vue'
import Project from '@/models/project'
import { entityKeys, entityKeysMap, PaperEntityTableColumn } from '@/models/paperEntities'

jest.mock('debounce', () => {
  return {
    debounce: (fn: Function) => fn
  }
})

describe('components/project/explore/DataTable.vue', () => {
  const store = {
    getters: {
      'projects/activeProject': new Project({ id: 42, name: 'Awesome Project' }),
      'paperEntities/entities': [{ id: 1, title: 'First Record' }, { id: 2, title: 'Second Record' }],
      'paperEntities/entityCount': 12
    },
    dispatch: jest.fn()
  }
  const router = {
    push: jest.fn()
  }

  function getWrapper (type: string = 'record') {
    return shallowMount(DataTable, {
      i18n,
      propsData: {
        queryByType: type
      },
      mocks: {
        $store: store,
        $router: router,
        $route: {
          params: {
            projectId: 42,
            entityType: type
          }
        }
      }
    })
  }

  it('renders', () => {
    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('it renders the different entity types', () => {
    entityKeys.forEach((entityKey) => {
      describe(entityKey, () => {
        let wrapper: any
        beforeEach(() => {
          wrapper = getWrapper(entityKey)
        })
        it('renders the title', () => {
          const title = wrapper.find('h1')
          expect(title.exists()).toBeTruthy()
          expect(title.text()).toEqual('project.explore.title.query_by_type.' + entityKey)
        })

        it('renders the correct headers', () => {
          const headers = wrapper.vm.headers
          expect(headers.map((header: any) => ({
            text: header.text,
            value: header.value
          }))).toMatchSnapshot()
        })

        it('navigates to an entity on row click', () => {
          wrapper.vm.handleRowClick(store.getters['paperEntities/entities'][0])

          expect(router.push).toHaveBeenCalledWith({
            name: 'projects.single.explore.view',
            params: {
              projectId: 42,
              entityType: entityKey,
              entityId: 1
            }
          })
        })
      })
    })
  })

  describe('handles the reloading of the data', () => {
    let wrapper: any
    beforeEach(() => {
      wrapper = getWrapper()
      store.dispatch.mockClear()
    })
    it('reloads the data if the page options updates', () => {
      wrapper.vm.updateOptions({
        page: 2,
        itemsPerPage: 5,
        sortBy: [],
        sortDesc: [],
        searchBy: [],
        search: ''
      })

      expect(store.dispatch).toHaveBeenCalledWith('paperEntities/fetchEntities', {
        projectId: 42,
        entityType: 'record',
        filterBy: undefined,
        perPage: 5,
        pageOffset: 1,
        filter: ''
      })
    })
    it('calculates the correct sort parameter string', () => {
      wrapper.vm.updateOptions({
        page: 2,
        itemsPerPage: 5,
        sortBy: ['authors'],
        sortDesc: [true],
        searchBy: [],
        search: ''
      })

      expect(store.dispatch).toHaveBeenCalledWith('paperEntities/fetchEntities', {
        projectId: 42,
        entityType: 'record',
        perPage: 5,
        pageOffset: 1,
        filterBy: undefined,
        filter: '',
        sortBy: 'authors_DESC'
      })

      wrapper.vm.updateOptions({
        page: 2,
        itemsPerPage: 5,
        sortBy: ['title'],
        sortDesc: [false],
        search: ''
      })

      expect(store.dispatch).toHaveBeenCalledWith('paperEntities/fetchEntities', {
        projectId: 42,
        entityType: 'record',
        perPage: 5,
        pageOffset: 1,
        filter: '',
        sortBy: 'title_ASC'
      })
    })
    it('prevents the reload data if the page options stays the same', () => {
      wrapper.vm.updateOptions({
        page: 1,
        itemsPerPage: 10,
        sortBy: [],
        sortDesc: [],
        search: '',
        searchBy: []
      })

      expect(store.dispatch).not.toHaveBeenCalled()
    })

    it('reloads the data if the search updates', () => {
      wrapper.vm.currentOptions = {
        page: 2,
        itemsPerPage: 5,
        sortBy: [],
        sortDesc: [],
        search: ''
      }
      wrapper.vm.handleSearch('my test')

      expect(store.dispatch).toHaveBeenCalledWith('paperEntities/fetchEntities', {
        projectId: 42,
        entityType: 'record',
        perPage: 5,
        pageOffset: 0,
        filter: 'my test'
      })

      expect(wrapper.vm.currentOptions).toEqual({
        page: 1,
        itemsPerPage: 5,
        sortBy: [],
        sortDesc: [],
        search: 'my test'
      })
    })
    it('prevents the reload data if the search stays the same', () => {
      wrapper.vm.handleSearch('', '')

      expect(store.dispatch).not.toHaveBeenCalled()

      wrapper.vm.currentOptions.search = 'test'

      store.dispatch.mockClear()

      wrapper.vm.handleSearch('test', 'test')

      expect(store.dispatch).not.toHaveBeenCalled()
    })
  })

  it('handles a change of the type', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    wrapper.vm.currentOptions = {
      page: 2,
      itemsPerPage: 5,
      sortBy: ['title'],
      sortDesc: [true],
      search: 'test'
    }

    // @ts-ignore
    wrapper.vm.updateQueryByType()
    // @ts-ignore
    expect(wrapper.vm.currentOptions).toEqual({
      page: 1,
      itemsPerPage: 10,
      sortBy: [],
      sortDesc: [],
      searchBy: [],
      search: ''
    })
    expect(store.dispatch).toHaveBeenCalledWith('paperEntities/fetchEntities', {
      projectId: 42,
      entityType: 'record',
      perPage: 10,
      pageOffset: 0,
      filter: ''
    })
  })

  it('provides a function to nicely display all authors names', () => {
    const wrapper = getWrapper()
    const item = {
      authors: [
        { firstName: 'Hans', lastName: 'Meier' },
        { firstName: 'Max', lastName: 'Dorfer' }
      ]
    }
    // @ts-ignore
    expect(wrapper.vm.getAuthorNames(item)).toEqual(['H. Meier', 'M. Dorfer'])
    // @ts-ignore
    expect(wrapper.vm.getAuthorNames({})).toEqual([])
  })

  it('provides a function to nicely display an authors name', () => {
    const wrapper = getWrapper()
    const author = { firstName: 'Hans', lastName: 'Meier' }
    // @ts-ignore
    expect(wrapper.vm.getAuthorDisplayName(author)).toEqual('H. Meier')
  })

  it('display no items if no headers are selected', () => {
    const backUp = entityKeysMap['record']
    // @ts-ignore
    Object.defineProperty(entityKeysMap, 'record', {
      value: {
        displayedColumns: []
      }
    })

    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
    // @ts-ignore
    expect(wrapper.vm.headers).toEqual([])
    // @ts-ignore
    expect(wrapper.vm.items).toEqual([])

    // @ts-ignore
    Object.defineProperty(entityKeysMap, 'record', {
      value: backUp
    })
  })

  it('provides the search options', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    const options = wrapper.vm.searchByOptions as Array<PaperEntityTableColumn>
    expect.assertions(options.length + 2)
    expect(options).toMatchSnapshot()
    expect(options).toEqual(expect.any(Array))
    options.forEach((option) => {
      expect(option.filterable).toBeTruthy()
    })
  })

  it('can render a title', () => {
    const wrapper = getWrapper()
    wrapper.setProps({
      title: 'This is my title'
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('This is my title')
  })
})
