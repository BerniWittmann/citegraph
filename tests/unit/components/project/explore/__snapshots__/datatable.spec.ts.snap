// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/project/explore/DataTable.vue it renders the different entity types author renders the correct headers 1`] = `
Array [
  Object {
    "text": "project.explore.table.headers.author.first_name",
    "value": "firstName",
  },
  Object {
    "text": "project.explore.table.headers.author.last_name",
    "value": "lastName",
  },
  Object {
    "text": "project.explore.table.headers.author.count_records",
    "value": "countRecords",
  },
]
`;

exports[`components/project/explore/DataTable.vue it renders the different entity types conference renders the correct headers 1`] = `Array []`;

exports[`components/project/explore/DataTable.vue it renders the different entity types country renders the correct headers 1`] = `Array []`;

exports[`components/project/explore/DataTable.vue it renders the different entity types institution renders the correct headers 1`] = `Array []`;

exports[`components/project/explore/DataTable.vue it renders the different entity types publication renders the correct headers 1`] = `Array []`;

exports[`components/project/explore/DataTable.vue it renders the different entity types publisher renders the correct headers 1`] = `Array []`;

exports[`components/project/explore/DataTable.vue it renders the different entity types record renders the correct headers 1`] = `
Array [
  Object {
    "text": "project.explore.table.headers.record.title",
    "value": "title",
  },
  Object {
    "text": "project.explore.table.headers.record.authors",
    "value": "renderedAuthors",
  },
  Object {
    "text": "project.explore.table.headers.record.year",
    "value": "year",
  },
  Object {
    "text": "project.explore.table.headers.record.number_citations",
    "value": "numberCitations",
  },
  Object {
    "text": "project.explore.table.headers.record.keywords",
    "value": "renderedKeywords",
  },
]
`;

exports[`components/project/explore/DataTable.vue renders 1`] = `
<v-card-stub loaderheight="4" tag="div" style="width: 100%;">
  <v-card-title-stub>
    <h1 class="display-1 my-5">
      project.explore.title.query_by_type.record
    </h1>
    <v-spacer-stub></v-spacer-stub>
    <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" appendicon="mdi-magnify" backgroundcolor="" hidedetails="true" label="project.explore.table.search" loaderheight="2" clearable="true" clearicon="$clear" singleline="true" type="text"></v-text-field-stub>
  </v-card-title-stub>
  <v-data-table-stub items="[object Object],[object Object]" options="[object Object]" sortby="" sortdesc="" customsort="function sortItems(items, sortBy, sortDesc, locale, customSorters) {
  if (sortBy === null || !sortBy.length) return items;
  var numericCollator = new Intl.Collator(locale, {
    numeric: true,
    usage: 'sort'
  });
  var stringCollator = new Intl.Collator(locale, {
    sensitivity: 'accent',
    usage: 'sort'
  });
  return items.sort(function (a, b) {
    var _a, _b;

    for (var i = 0; i < sortBy.length; i++) {
      var sortKey = sortBy[i];
      var sortA = getObjectValueByPath(a, sortKey);
      var sortB = getObjectValueByPath(b, sortKey);

      if (sortDesc[i]) {
        _a = __read([sortB, sortA], 2), sortA = _a[0], sortB = _a[1];
      }

      if (customSorters &amp;&amp; customSorters[sortKey]) {
        var customResult = customSorters[sortKey](sortA, sortB);
        if (!customResult) continue;
        return customResult;
      } // Check if both cannot be evaluated


      if (sortA === null &amp;&amp; sortB === null) {
        continue;
      }

      _b = __read([sortA, sortB].map(function (s) {
        return (s || '').toString().toLocaleLowerCase();
      }), 2), sortA = _b[0], sortB = _b[1];

      if (sortA !== sortB) {
        if (!isNaN(sortA) &amp;&amp; !isNaN(sortB)) return numericCollator.compare(sortA, sortB);
        return stringCollator.compare(sortA, sortB);
      }
    }

    return 0;
  });
}" page="1" itemsperpage="10" groupby="" groupdesc="" locale="en-US" disablefiltering="true" customfilter="function defaultFilter(value, search, item) {
  return value != null &amp;&amp; search != null &amp;&amp; typeof value !== 'boolean' &amp;&amp; value.toString().toLocaleLowerCase().indexOf(search.toLocaleLowerCase()) !== -1;
}" serveritemslength="12" itemkey="id" value="" expanded="" noresultstext="$vuetify.dataIterator.noResultsText" nodatatext="$vuetify.noDataText" loadingtext="$vuetify.dataIterator.loadingText" headers="[object Object],[object Object],[object Object],[object Object],[object Object]" mobilebreakpoint="600" expandicon="$expand" style="width: 100%;"></v-data-table-stub>
</v-card-stub>
`;
